src/core/controllers.ts

```diff
@@ -0,0 +1,156 @@
+import debug from 'debug'
+import { RequestHandler, Dictionary } from 'express-serve-static-core'
+import { AccountServices, SettlementEngine } from '.'
+import { isSafeKey, SafeKey, SettlementStore } from './store'
+import { fromQuantity, isQuantity } from './utils/quantity'
+import uuid from 'uuid/v4'
+import BigNumber from 'bignumber.js'
+
+const log = debug('settlement-core')
+
+export interface Context {
+  services: AccountServices
+  engine: SettlementEngine
+  store: SettlementStore
+}
+
+interface AccountParams extends Dictionary<string> {
+  id: SafeKey
+}
+
+interface SettlementController {
+  setupAccount: RequestHandler
+  validateAccount: RequestHandler
+  settleAccount: RequestHandler<AccountParams>
+  handleMessage: RequestHandler<AccountParams>
+  deleteAccount: RequestHandler<AccountParams>
+}
+
+export const createController = ({ store, engine, services }: Context): SettlementController => ({
+  setupAccount: async (req, res) => {
+    const accountId = req.body.id || uuid() // Create account ID if none was provided
+    if (!isSafeKey(accountId)) {
+      return res.status(400).send('Account ID includes unsafe characters')
+    }
+
+    /**
+     * TODO Before creating the accout/calling setup, ensure the peer is reachable:
+     * Try pinging them and await either a response or a ping from said peer
+     */
+
+    try {
+      await store.createAccount(accountId)
+    } catch (err) {
+      log(`Failed to setup account: account=${accountId}`, err)
+      return res.sendStatus(500)
+    }
+
+    if (engine.setupAccount) {
+      try {
+        await engine.setupAccount(accountId) // TODO Is it safe if this is called multiple times?
+      } catch (err) {
+        log(`Failed to setup account: account=${accountId}`, err)
+        return res.sendStatus(500)
+      }
+    }
+
+    res.status(201).send({
+      id: accountId
+    })
+  },
+
+  validateAccount: async (req, res, next) => {
```

Comment:
Calling this and then creating the account leaves open the possibility of a race condition. Should there be an atomic operation to add the account if it doesn't already exist instead?

Reply:
The `validateAccount` middleware isn't used when creating an account (https://github.com/interledgerjs/settlement-core/blob/master/src/index.ts#L331). The current behavior is to allow "create account" to be called multiple times without an error